### PR TITLE
🐛 [Story video] Manage bitrate on pool elements only (from amp video)

### LINKED ISSUE #32597: [Story video] Avoid managing placeholders on amp-video with mediapool
When an `amp-video` is managed by a mediapool, it shouldn't make the BitrateManager manage `this.video_`. It should only manage the mediapool element. This will improve performance and remove potential sources of errors while waiting for operations on the wrong element.

Look at:
- `amp-video`: Prevent calling `getBitrateManager(this.win).manage(this.video_)` on the placeholder if managed by a pool.
- `media-pool`: Stop managing/unmanaging the placeholder element on `swapElements` functions.

@ampproject/wg-stories

### PR DESCRIPTION
**Pull request by mszylkowski**
When a video is swapped in (in `resetOnDomChange`), manage the bitrate.
Do not manage bitrate on placeholder elements.

Closes #32597

### CODE REVIEW

File: extensions/amp-story/1.0/amp-story-page.js [hunk 1]
```diff
@@ -1040,9 +1040,9 @@ export class AmpStoryPage extends AMP.BaseElement {
    * @private
    */
   preloadAllMedia_() {
-    return this.whenAllMediaElements_((mediaPool, mediaEl) => {
-      this.preloadMedia_(mediaPool, mediaEl);
-    });
+    return this.whenAllMediaElements_((mediaPool, mediaEl) =>
+      this.preloadMedia_(mediaPool, mediaEl)
+    );
```
Comment (mszylkowski) [inline extensions/amp-story/1.0/amp-story-page.js:1045]: Before it was not waiting for the promise, now with a lambda function it returns the promise so it waits for it.

File: extensions/amp-story/1.0/media-pool.js [hunk 2]
```diff
@@ -514,25 +514,31 @@ export class MediaPool {
...
-        this.forceDeallocateMediaElement_(poolMediaEl);
-      }
-    );
+    )
+      .then(() =>
+        Promise.all([
+          this.maybeResetAmpMedia_(ampMediaForPoolEl),
+          this.maybeResetAmpMedia_(ampMediaForDomEl),
+        ])
+      )
```
Comment (mszylkowski) [inline extensions/amp-story/1.0/media-pool.js:523]: Wait for bitrate management before swapping sources

File: extensions/amp-story/1.0/media-pool.js [hunk 3]
```diff
@@ -514,25 +514,31 @@ export class MediaPool {
...
+      .then(
+        () => {
+          const sourcesPromise = this.enqueueMediaElementTask_(
```
Comment (mszylkowski) [inline extensions/amp-story/1.0/media-pool.js:526]: Return sources so `playAllMedia` can play videos when they have the sources. Otherwise `loadPromise` would fail.

File: extensions/amp-story/1.0/media-pool.js [hunk 4]
```diff
@@ -545,7 +551,7 @@ export class MediaPool {
       return;
     }
 
-    componentEl.getImpl().then((impl) => {
+    return componentEl.getImpl().then((impl) => {
```
Comment (mszylkowski) [inline extensions/amp-story/1.0/media-pool.js:554]: Make `maybeResetAmpMedia_` async so the `swap..MediaIntoDom_` can wait for this.

File: extensions/amp-video/0.1/amp-video.js [hunk 5]
```diff
@@ -168,6 +168,9 @@ export class AmpVideo extends AMP.BaseElement {
     /** @private {boolean} */
     this.isPlaying_ = false;
 
+    /** @private {boolean} */
+    this.isManagedByBitrate_ = this.hasAnyBitrateSources_();
```
Comment (mszylkowski) [inline extensions/amp-video/0.1/amp-video.js:172]: It's easier if it's cached, and avoids calculating it when in between mediapool operations, which could return the wrong value (ex: in between SwapIntoDom and UpdateSources)

File: extensions/amp-video/0.1/amp-video.js [hunk 6]
```diff
@@ -575,6 +576,33 @@ export class AmpVideo extends AMP.BaseElement {
...
 
+  /**
+   * @private
+   * @return {boolean}
+   */
+  hasAnyBitrateSources_() {
+    return (
+      !!this.element.querySelector('source[data-bitrate]') ||
+      this.hasAnyCachedSources_()
+    );
```
Comment (mszylkowski) [inline extensions/amp-video/0.1/amp-video.js:587]: If it has cachedSources it will generate the bitrate sources in the buildCallback, so we can assume it has bitrate sources if it has cached sources

File: extensions/amp-video/0.1/amp-video.js [hunk 7]
```diff
@@ -632,11 +660,15 @@ export class AmpVideo extends AMP.BaseElement {
...
     );
-
     this.uninstallEventHandlers_();
     this.installEventHandlers_();
+    if (this.isManagedByBitrate_) {
+      getBitrateManager(this.win).manage(this.video_);
+    }
     // When source changes, video needs to trigger loaded again.
-    this.loadPromise(this.video_).then(() => this.onVideoLoaded_());
+    listenOncePromise(this.video_, 'loadedmetadata').then(() =>
```
Comment (mszylkowski) [inline extensions/amp-video/0.1/amp-video.js:669]: `this.loadPromise` would not work here since it won't have the sources attached yet, so we use the loadedMetadata directly here.

File: extensions/amp-video/0.1/amp-video.js [hunk 8]
```diff
@@ -768,6 +800,18 @@ export class AmpVideo extends AMP.BaseElement {
...
+   * Whether the video is just a placeholder and meant to be overriden
+   * @param {!Element} videoEl
+   * @return {boolean}
+   */
+  isPlaceholderVideo_(videoEl) {
+    if (!this.isManagedByPool_()) {
+      return false;
+    }
+    return !videoEl.classList.contains('i-amphtml-pool-media');
+  }
```
Comment (mszylkowski) [inline extensions/amp-video/0.1/amp-video.js:813]: @gmajoulet We may not need this since we only check that on buildCallback, where the mediapool probably didn't swap the video yet. If that's the case, we can remove this function.

File: extensions/amp-video/0.1/flexible-bitrate.js [hunk 9]
```diff
@@ -91,6 +91,9 @@ export class BitrateManager {
    * @param {!Element} video
    */
   manage(video) {
+    if (video.changedSources) {
+      return;
+    }
```
Comment (mszylkowski) [inline extensions/amp-video/0.1/flexible-bitrate.js:96]: If mediapoolEl was already managed, no need to redo this. It will call `changedSources` by itself on the `UpdateSourcesTask`

**Review by mszylkowski — COMMENTED** [review]
@gmajoulet @newmuis Updated code and wrote extensive comments on this PR to follow up more easily. PTAL

File: extensions/amp-story/1.0/media-pool.js (hunk 3 above)
Comment (gmajoulet) [reply extensions/amp-story/1.0/media-pool.js:526]: let's write this code in the most predictable way and wait for all actions to be done before its promise resolves:

```
.then(() => enqueue UpdateSourcesTask)
.then(() => enqueue LoadTask)
```

File: extensions/amp-story/1.0/media-pool.js (hunk 4 above)
Comment (gmajoulet) [reply extensions/amp-story/1.0/media-pool.js:554]: It's pretty cool that `Promise.all` handles `undefined` so well but let's change this method to be `@return {!Promise}` it'll make things easier

File: extensions/amp-video/0.1/amp-video.js (hunk 7 above)
Comment (gmajoulet) [reply extensions/amp-video/0.1/amp-video.js:669]: Can you please add a code comment? "Video might not have its sources yet"...

Comment (gmajoulet) [reply extensions/amp-video/0.1/amp-video.js:669]: You're only handling the case where the video didn't start loading. If the load has started this will never resolve as the `loadedmetadata` event was already fired.

You need to add an extra check on the video readyState:
- If readyState >= 1, onVideoLoaded_()
- otherwise when loadedmetadata, onVideoLoaded()

File: extensions/amp-video/0.1/amp-video.js (hunk 8 above)
Comment (gmajoulet) [reply extensions/amp-video/0.1/amp-video.js:813]: The media pool waits for the video build/layoutCallbacks to resolve, you can safely remove this check.

File: extensions/amp-video/0.1/amp-video.js [hunk 10]
```diff
@@ -222,9 +225,7 @@ export class AmpVideo extends AMP.BaseElement {
 
     this.video_ = element.ownerDocument.createElement('video');
     // Manage video if the sources contain bitrate or amp-orig-src will be expanded to multiple bitrates.
-    if (
-      this.element.querySelector('source[data-bitrate], source[amp-orig-src]')
-    ) {
+    if (this.isManagedByBitrate_ && !this.isPlaceholderVideo_(this.video_)) {
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/amp-video.js:228]: Are these changes required for this PR to work? If not, let's remove them for now

Comment (mszylkowski) [reply extensions/amp-video/0.1/amp-video.js:228]: This is because we don't want to manage the placeholder video, only the mediapool video. Not sure if removing this change is breaking, but it's the main point of this PR's idea to move the management from here to the `resetOnDomChange`

File: extensions/amp-video/0.1/amp-video.js (hunk 7 above)
Comment (mszylkowski) [reply extensions/amp-video/0.1/amp-video.js:669]: This is for preemptively safety right? I don't think there's a way that readyState can be >=1 here since the `UpdateSourcesTask` is called after this method and we fixed preload Promise.

File: extensions/amp-video/0.1/amp-video.js (hunk 8 above)
Comment (mszylkowski) [reply extensions/amp-video/0.1/amp-video.js:813]: SG, then this can be replaced with `this.isManagedByPool()`

File: extensions/amp-video/0.1/amp-video.js [hunk 11]
```diff
@@ -222,9 +225,7 @@ export class AmpVideo extends AMP.BaseElement {
 
     this.video_ = element.ownerDocument.createElement('video');
     // Manage video if the sources contain bitrate or amp-orig-src will be expanded to multiple bitrates.
-    if (
-      this.element.querySelector('source[data-bitrate], source[amp-orig-src]')
-    ) {
+    if (this.isManagedByBitrate_ && !this.isManagedByPool_()) {
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/amp-video.js:228]: I think we should just remove this code, only Stories use the ABR, let's not create dead paths.

File: extensions/amp-video/0.1/amp-video.js (hunk 5 above)
Comment (gmajoulet) [reply extensions/amp-video/0.1/amp-video.js:172]: Okay, let's:

- move this work to the `buildCallback` (media pool will wait until build/layoutCallback resolves so it's safe)
- rename to something even more accurate about what it's used for "should be managed by the flexible bitrate class", rn it sounds like it's already managed and "bitrate" is a bit unclear/misleading

File: extensions/amp-video/0.1/flexible-bitrate.js (hunk 9 above)
Comment (gmajoulet) [reply extensions/amp-video/0.1/flexible-bitrate.js:96]: plz add a code comment `if already managed.....`

File: extensions/amp-video/0.1/amp-video.js (hunk 5 above)
Comment (gmajoulet) [reply extensions/amp-video/0.1/amp-video.js:172]: Maybe also add a code comment in the buildCallback to say that we look at the sources here before the mediapool removes them

**Review by gmajoulet — COMMENTED** [review]
Some nits, but this looks good!

File: extensions/amp-video/0.1/amp-video.js (hunk 11 above)
Comment (mszylkowski) [reply extensions/amp-video/0.1/amp-video.js:228]: Ok SG 👍

File: extensions/amp-video/0.1/amp-video.js [hunk 12]
```diff
@@ -262,6 +259,11 @@ export class AmpVideo extends AMP.BaseElement {
       'artwork': [{'src': artwork || poster || ''}],
     };
 
+    // Cached so mediapool operations (eg: swapping sources) don't interfere with this bool.
+    this.hasBitrateSources_ =
+      !!this.element.querySelector('source[data-bitrate]') ||
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/amp-video.js:264]: Up to you, but we might want an early return if (mediaPool) so we have 0 impact on non stories doc

**Review by gmajoulet — APPROVED** [review]
embarcación it 🚢

File: extensions/amp-video/0.1/amp-video.js (hunk 12 above)
Comment (mszylkowski) [reply extensions/amp-video/0.1/amp-video.js:264]: Ok I see what you mean. We can have an early return if we end up doing more here, but this variable is only used on `resetOnDomChange` so it already has 0 impact on non-stories docs. Let's keep it to the simplest code possible so it's easy to understand and debug, but if we need to shortcircuit the expression at some point, we can do it.
